### PR TITLE
[Networking] Caching application specific score of GossipSub

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR offers two significant enhancements:
- First, it introduces a caching mechanism for the GossipSub's application-specific score. This mechanism notably reduces the cache read operations for each incoming message by two-thirds, decreasing from three reads to just one. This reduction is a key part of optimizing [resource-intensive RPC inspection operations](https://github.com/dapperlabs/flow-go/issues/6870). 
- Second, the PR undertakes a comprehensive refactoring of the networking layer's configuration structure. It enhances the coherence, integration, and organization of the configuration settings, creating a streamlined, top-down approach from the configuration file right through to the individual components of the networking layer.

app-specific-score-cache

### CODE REVIEW

File: config/default-config.yml [hunk 1]
```diff
@@ -117,92 +117,123 @@ network-config:
...
-  gossipsub-app-specific-penalty-decay-rate-reduction-factor: .01
-  # Defines the frequency for evaluating and potentially adjusting the decay process of a spam record. At each interval, the system assesses the current penalty of a node. If this penalty is below the defined threshold, the decay rate is modified according to the reduction factor, slowing down the penalty reduction process. This reassessment at regular intervals ensures that the decay rate is dynamically adjusted to reflect the node's ongoing behavior, maintaining a balance between penalizing malicious activity and allowing recovery for honest nodes.
-  gossipsub-app-specific-penalty-decay-evaluation-period: 10m
+  gossipsub:
```
Comment (kc1116) [inline config/default-config.yml:120]: Nice !

File: network/p2p/scoring/registry.go [hunk 2]
```diff
@@ -170,66 +219,139 @@ func NewGossipSubAppSpecificScoreRegistry(config *GossipSubAppSpecificScoreRegis
...
+			submitted := r.appScoreUpdateWorkerPool.Submit(pid)
+			lg.Trace().
+				Bool("worker_submitted", submitted).
+				Float64("app_specific_score", appSpecificScore).
+				Dur("score_ttl", r.scoreTTL).
+				Msg("application specific score expired, submitting worker to update it")
+
+			return appSpecificScore // in the mean time, return the expired score.
+		default:
+			// record found in the cache, check if it is expired.
```
Comment (kc1116) [inline network/p2p/scoring/registry.go:266]:
```suggestion
			// record found in the cache.
```

File: utils/unittest/math.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package unittest
```
Comment (kc1116) [inline utils/unittest/math.go:1]: 🙌

**Review by kc1116 — APPROVED** [review]
🚀

File: network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go [hunk 4]
```diff
@@ -282,25 +248,24 @@ func defaultInspectorSuite(rpcTracker p2p.RpcControlTracking) p2p.GossipSubRpcIn
 	}
 }
 
-// Build creates a new GossipSub pubsub system.
-// It returns the newly created GossipSub pubsub system and any errors encountered during its creation.
+// Build creates a new GossipSubParameters pubsub system.
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go]: thanks for catching that, fixed https://github.com/onflow/flow-go/pull/5045/commits/9c96d156dc4214fa73be2cfb88158ac26b2b23ef

File: network/p2p/scoring/internal/appSpecificScoreCache.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+			return fmt.Errorf("failed to add app specific score record for peer %s", peerID)
+		}
+
+		u := updated.(appSpecificScoreRecordEntity)
+		if u.Score != score {
+			return fmt.Errorf("incorrect update of app specific score record for peer %s, expected score %f, got score %f", peerID, score, u.Score)
+		}
+		if u.LastUpdated != time {
+			return fmt.Errorf("incorrect update of app specific score record for peer %s, expected last updated %s, got last updated %s", peerID, time, u.LastUpdated)
+		}
```
Comment (peterargue) [inline network/p2p/scoring/internal/appSpecificScoreCache.go:99]: what are these for?

Comment (yhassanzadeh13) [reply network/p2p/scoring/internal/appSpecificScoreCache.go:99]: internal sanity check for the cache, to fail fast if needed.

Comment (peterargue) [reply network/p2p/scoring/internal/appSpecificScoreCache.go:99]: I guess I'm just not clear on why it's needed. The update done within `Adjust` is atomic. Do you not trust the implementation, or is there a corner case you're worried about?

Comment (yhassanzadeh13) [reply network/p2p/scoring/internal/appSpecificScoreCache.go]: Your observation is valid; the sanity check is redundant given that `Adjust` is designed to be atomic. I have eliminated this check as reflected (https://github.com/onflow/flow-go/pull/5045/commits/669af4ef3cc59525090d186ba862ac8f4bb9eb33). Initially, my rationale was to implement a sanity check for updates related to scoring, considering the complexities encountered while debugging live scoring issues. However, given the comprehensive testing we've implemented in the current PR, the probability of encountering edge cases is significantly reduced. As a result, this sanity check appears more like a redundant safeguard and has been removed accordingly.

**Review by peterargue — APPROVED** [review]

**Pull request merged**